File: 2024/two/src/main.rs
//! Advent of Code day 2

use std::fs;

fn is_safe(report: &Vec<i32>) -> bool {
    let mut is_safe = true;
    let mut prev_diff = 0i32;
    for ctr in 0..(report.len()-1) {
        let diff = report[ctr+1] - report[ctr];
        let abs_diff = diff.abs();
        if abs_diff < 1 || abs_diff > 3 {
            is_safe = false;
            break;
        }
        if ctr == 0 {
            prev_diff = diff;
            continue;
        }
        if prev_diff * diff < 0 {
            return false;
        }
        prev_diff = diff;
    }
    return is_safe;
}

fn main() {
    let mut safe_pure: u32 = 0;
    let mut unsafe_pure: u32 = 0;

    let mut safe_dampened: u32 = 0;
    let mut unsafe_dampened: u32 = 0;
    
    for line in fs::read_to_string("./input").unwrap().lines() {
        // Parse a single report
        let report: Vec<i32> = line
                                .trim()
                                .split(" ")
                                .map(|x| x.parse::<i32>().unwrap())
                                .collect();

        // Part 1: Without problem dampener
        // Check for safe and unsafe reports
        if is_safe(&report) {
            safe_pure += 1;
        } else {
            unsafe_pure += 1;
        }

        // Part 2: With problem dampener
        // Check for safe and unsafe reports
        if is_safe(&report) {
            safe_dampened += 1;
        } else {
            let mut safe: bool = false;
            for idx in 0..report.len() {
                let mut copied = report.clone();
                copied.remove(idx);
                if is_safe(&copied) {
                    safe = true;
                    break;
                }
            }
            if safe {
                safe_dampened += 1;
            } else {
                unsafe_dampened += 1;
            }
        }
    }
    println!("Part 1: safe: {}, unsafe: {}", &safe_pure, &unsafe_pure);
    println!("Part 2: safe: {}, unsafe: {}", &safe_dampened, &unsafe_dampened);
}

File: 2024/two/main.py
"""
Advent of code day two
"""
from copy import deepcopy

Report = list[int]

def is_safe(report: Report) -> bool:
    is_safe = True
    prev_diff = 0
    for ctr in range(len(report)-1):
        diff = report[ctr+1] - report[ctr]
        if not 1 <= abs(diff) <= 3:
            is_safe = False
        if ctr == 0:
            prev_diff = diff
            continue
        if prev_diff * diff < 0:
            return False
        prev_diff = diff
    return is_safe

def check_pure(reports: list[Report]) -> tuple[int, int, list[Report], list[Report]]:
    safe, unsafe = 0, 0
    safe_reports, unsafe_reports = [], []
    for report in reports:
        if is_safe(report):
            safe_reports.append(report)
            safe += 1
        else:
            unsafe_reports.append(report)
            unsafe += 1
    print(f"{safe=}, {unsafe=}")
    return safe, unsafe, safe_reports, unsafe_reports

def check_dampened(reports: list[Report]) -> tuple[int, int, list[Report], list[Report]]:
    safe, unsafe = 0, 0
    safe_reports, unsafe_reports = [], []
    for report in reports:
        if is_safe(report):
            safe_reports.append(report)
            safe += 1
        else:
            safe_cond = False
            for idx in range(len(report)):
                copied = deepcopy(report)
                copied.pop(idx)
                if is_safe(copied):
                    safe_cond = True
                    break
            if safe_cond:
                safe_reports.append(report)
                safe += 1
            else:
                unsafe_reports.append(report)
                unsafe += 1
    print(f"{safe=}, {unsafe=}")
    return safe, unsafe, safe_reports, unsafe_reports


def main():
    with open("./input", "r") as f:
        reports = [list(map(lambda x: int(x), line.strip().split(" "))) for line in f.readlines()]
    print("*********Checking pure**************")
    safe_pure, unsafe_pure, safe_reports_pure, unsafe_reports_pure = check_pure(reports)
    print("************************************")
    print("*********Checking dampened**********")
    safe_dampened, unsafe_dampened, safe_reports_dampened, unsafe_reports_dampened = check_dampened(reports)
    print("************************************")

if __name__=="__main__":
    main()

